### PR TITLE
Setup google authentication

### PR DESCRIPTION
**Pull request by EGiataganas**
### Description of change

- Update package to include `cookie` and `jsonwebtoken`
- Setup Google authentication
- Set environment key using parameter store
- Tweak branch to test auth
- Update and fix tests to use gsso cookie
- Update readme
- Add authorisation check on pages api
- Move AuthHeader to properties
- Show logout link only when user is logged in

### Story Link

https://trello.com/c/kHrLO5Y0

### CODE REVIEW

File: src/components/Layout/Header/Header.test.js [hunk 1]
```diff
@@ -1,11 +1,36 @@
...
-    const { getByText } = render(<Header serviceName={serviceName} />)
-    const header = getByText(serviceName)
-    expect(header).toBeInTheDocument()
+  const serviceName = 'Hackney Header'
+
+  it('should render service name with logout link', () => {
+    const { getByText } = render(
+      <UserContext.Provider
+        value={{
+          user: { name: 'bar' },
```
Comment (NeilvB) [inline src/components/Layout/Header/Header.test.js:12]: Unless we need the name for something, I suggest we just leave this as an empty object

File: cypress/integration/home_page.spec.js [hunk 2]
```diff
@@ -3,9 +3,18 @@
...
+  beforeEach(() => {
+    cy.getCookies().should('be.empty')
+    cy.setCookie('hackneyToken', Cypress.env('GSSO_TEST_KEY'))
+    cy.getCookie('hackneyToken').should(
+      'have.property',
+      'value',
+      Cypress.env('GSSO_TEST_KEY')
+    )
+    cy.visit(Cypress.env('HOST'))
+  })
```
Comment (benbaumann95) [inline cypress/integration/home_page.spec.js:15]: Hey as mentioned (for another PR anyway) we can put this into a helper function as all tests will probably need to use this

File: src/pages/_app.js [hunk 3]
```diff
@@ -3,26 +3,39 @@ import App from 'next/app'
...
+MyApp.getInitialProps = async ({ ctx }) => {
+  const WITH_REDIRECT = true
+
+  if (AUTH_WHITELIST.includes(ctx.pathname)) {
+    return {}
+  }
+
+  const userDetails = isAuthorised(ctx, WITH_REDIRECT)
+
+  if (NEXT_PUBLIC_ALLOWED_IP_ADDRESSES) {
```
Comment (benbaumann95) [inline src/pages/_app.js:38]: Do we need this check anymore ? Maybe can remove in next PR as part of your refactor/tests?

**Comment by benbaumann95** [comment]
Just going to test on staging before approving but looks great, nice job!

File: src/components/Layout/Header/Header.js [hunk 4]
```diff
@@ -1,15 +1,23 @@
...
+  const { user } = useContext(UserContext)
+
+  return (
+    <>
+      <Header
+        serviceName={serviceName}
+        isServiceNameShort={false}
+        homepageUrl="/"
+      >
+        {user && <a href="/logout">Logout</a>}
```
Comment (benbaumann95) [inline src/components/Layout/Header/Header.js:16]: Think we can/should update the UI of this too but think you already mentioned this

**Review by benbaumann95 — APPROVED** [review]

**Comment by benbaumann95** [comment]
Works fine on staging for me. 🚀 
Just remember to delete the .circleci/config.yml stuff but good to go i hope!

File: src/pages/_app.js (hunk 3 above)
Comment (EGiataganas) [reply src/pages/_app.js:38]: no we don't but I preferred to leave it on another PR

File: src/components/Layout/Header/Header.test.js (hunk 1 above)
Comment (EGiataganas) [reply src/components/Layout/Header/Header.test.js:12]: I thought is better to have the name there as we're having it. But happy to leave it as an empty object

**Pull request merged**